### PR TITLE
Add `quicktoctree` directive for easier prioritised globbed ToC reuse

### LINKED ISSUE #1861: Explore fine-grained manual reorganisation of documentation sidebar items
## Problem

Furo, the Sphinx theme we use, organises sidebar content alphabetically based on directory name. We do not want to use artifically named directories to force a specific ordering, but we do want to have logically grouped content in the sidebar.

## Description

Use a template override or small Sphinx extension to patch Furo's sorting behaviour to use our desired order.

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Fixes

Fixes #1861 by @sarayourfriend

## Description

I'm not sure if this is the best solution. However, it's an interesting one, at least. The alternative is really to copy and paste the same `toctree` options in all the places that use `quicktoctree` now.

This fixes the ordering problem raised in the issue without reintroducing the need to duplicate the `toctree` settings everywhere. However, it does so by introducing a custom directive that just extends the base `TocTree` directive. Maybe that's _too_ clever and we really should just copy/paste those `toctree` settings everywhere.

## Testing Instructions

Try out the new `quicktoctree` directive and confirm it works identically to `toctree` except for the new `type` option which adds the new behaviours.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [N/A] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [N/A] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: documentation/api/guides/index.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Guides
 
-```{include} ../../_toc_templates/flat.md
```
Comment (dhruvkb) [inline documentation/api/guides/index.md:3]: This isn't directly relevant but why do we use `_toc_templates` and not define the `toctree` manually to list content in a logical order?

**Review by dhruvkb — COMMENTED** [review]
Not an actual review, I have a question.

**Review by AetherUnbound — APPROVED** [review]
This looks good! It doesn't seem like that much of a duplication based on what we had before.

File: documentation/api/guides/index.md (hunk 1 above)
Comment (sarayourfriend) [reply documentation/api/guides/index.md:3]: We certainly could, I just found it tedious for sections where there was no clear reason to order things in any particular way. People frequently forget to add pages to the toctree directives[^note], so I think globs are nice in that way. Repeating the `:titlesonly:` and other options was tedious to me, so I tried originally to use these include directives, but they're not properly flexible.

Anyway, I'm not convinced that the approach in this PR is better than doing as you said, just writing out the toctrees individually. I just found it really tedious to repeat all the options for the two common "types" of toctrees we use.

[^note]: Seriously, I've had to fix so many "missing from toctree" warnings.

**Comment by sarayourfriend** [comment]
@krysal Can you give this one a review soon? :pray:

**Review by obulat — CHANGES_REQUESTED** [review]
I receive several warnings in the console when building the docs:
```
/openverse/openverse/documentation/api/index.md:7: WARNING: toctree glob pattern '*' didn't match any documents
/openverse/openverse/documentation/catalog/index.md:3: WARNING: toctree glob pattern '*' didn't match any documents
/openverse/openverse/documentation/catalog/index.md:3: WARNING: toctree glob pattern '*/index' didn't match any documents
/openverse/openverse/documentation/changelogs/frontend/index.md:5: WARNING: toctree glob pattern '*' didn't match any documents
/openverse/openverse/documentation/changelogs/index.md:11: WARNING: toctree glob pattern '*' didn't match any documents
/openverse/openverse/documentation/changelogs/ingestion_server/index.md:5: WARNING: toctree glob pattern '*' didn't match any documents
/openverse/openverse/documentation/frontend/index.md:3: WARNING: toctree glob pattern '*' didn't match any documents
/openverse/openverse/documentation/ingestion_server/index.md:3: WARNING: toctree glob pattern '*' didn't match any documents
```
Would moving the markdown documents from nested folders into the root folder help?
This is the main blocker for me for this PR.

Should these be fixed in a different PR?
```
/openverse/openverse/documentation/meta/ci_cd/jobs/ingestion_server.md:3: WARNING: Non-consecutive header level increase; H1 to H3 [myst.header]
/openverse/openverse/documentation/meta/ci_cd/jobs/preparation.md:3: WARNING: Non-consecutive header level increase; H1 to H3 [myst.header]
/openverse/openverse/documentation/meta/ci_cd/jobs/preparation.md:25: WARNING: Non-consecutive header level increase; H1 to H3 [myst.header]
/openverse/openverse/documentation/projects/proposals/document_all_media_properties/20230307-project_proposal.md:4: WARNING: Document headings start at H2, not H1 [myst.header]
/openverse/openverse/documentation/projects/proposals/document_all_media_properties/20230307-project_proposal.md:8: WARNING: Document headings start at H2, not H1 [myst.header]
/openverse/openverse/documentation/projects/proposals/document_all_media_properties/20230307-project_proposal.md:13: WARNING: Document headings start at H2, not H1 [myst.header]
```

**Comment by sarayourfriend** [comment]
> Should these be fixed in a different PR?

Yes, those specific errors should be addressed in a different PR as they're not caused by the toc tree issue. They're addressed I think in #855 and/or in #1973 (whichever is merged first).

The others are caused by those directories not including folders within themselves. I'm not sure how best to solve them without defeating the purpose of the `quicktoctree` "types". I'll do a bit of poking to try and find a workaround. Worst case, we can just use the regular `toctree` directive for those in the mean time.

> Can we still define toctree manually for places that require manual content ordering?

Yes, and `quicktoctree` still allows you to do so anyway. All it does is automatically add the globbed versions to prevent pages from not being included.

Though, now that I think about it, we could just close this PR and write the toctrees out manually. As annoying as it is to repeat those things, it's only a bulk job once. After that it should be rare outside adding new projects, when it's easy enough to do anyway. Adding new directories doesn't seem like a regular ocurrence and it's easy to handle when it is.

The main goal of this PR was to prevent documents from not being included. But in #1973 we prevent warnings from being committed anyway, so that'll never be an issue going forward after that PR.

I'm going to close this PR and update the toc trees in #1973 to fix this stuff once and for all without needing to write custom extensions :slightly_smiling_face: Thanks for challenging the approach @dhruvkb and @obulat!

**Pull request closed without merging**